fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java

```diff
@@ -0,0 +1,325 @@
+            resourceTypes = tmpResourceTypes.stream().filter(item-> allCompartmentResourceTypes.contains(item)).collect(Collectors.toList());
+            if (resourceTypes == null || resourceTypes.isEmpty()) {
+                throw new Exception("readItem: None of the input resource types is valid!");
+            }
+        }
+
+        if (fhirSearchPatientGroupId == null) {
+            throw new Exception("readItem: missing group id for this group export job!");
+        }
+
+        TransientUserData chunkData = (TransientUserData) jobContext.getTransientUserData();
+        if (chunkData != null) {
+            if (resourceTypes.size() == indexOfCurrentResourceType + 1) {
+                // No more resource type and page to read, so return null to end the reading.
+                return null;
+            } else {
+                // More resource types to read, so reset partNum and move resource type index to the next.
+                chunkData.setPartNum(1);
+                indexOfCurrentResourceType++;
+            }
+        }
+        if (fhirTenant == null) {
+            fhirTenant = Constants.DEFAULT_FHIR_TENANT;
+            logger.info("readItem: Set tenant to default!");
+        }
+        if (fhirDatastoreId == null) {
+            fhirDatastoreId = Constants.DEFAULT_FHIR_TENANT;
+            logger.info("readItem: Set DatastoreId to default!");
+        }
+        if (fhirSearchPageSize != null) {
+            try {
+                pageSize = Integer.parseInt(fhirSearchPageSize);
+                logger.fine("readItem: Set page size to " + pageSize + ".");
+            } catch (Exception e) {
+                logger.warning("readItem: Set page size to default(" + Constants.DEFAULT_SEARCH_PAGE_SIZE + ").");
+            }
+        }
+
+        Group group = FindGroupByID(fhirTenant, fhirDatastoreId, fhirSearchPatientGroupId);
+        // List for the patients
+        List<Member> patientMembers = new ArrayList<>();
+        // List for the group and sub groups in the expansion paths, this is used to avoid dead loop caused by circle reference of the groups.
+        HashSet<String> groupsInPath = new HashSet<>();
+        ExpandGroup2Patients(fhirTenant, fhirDatastoreId, group, patientMembers, groupsInPath);
+
+        if (chunkData == null) {
+            chunkData = new TransientUserData(0, null, new ArrayList<PartETag>(), 1);
+            chunkData.setIndexOfCurrentResourceType(0);
+            jobContext.setTransientUserData(chunkData);
+        } else {
+            chunkData.setIndexOfCurrentResourceType(indexOfCurrentResourceType);
+        }
+        // The fhir resources of one resource type for all the patients will be exported into one COS object.
+        // Here we simply set the lastPageNum to be smaller than the next PageNum to ask the common ChunkWriter
+        // to close the writing for current resource type.
+        chunkData.setPageNum(2);
```

Comment:
what does setPageNum 2 mean?

Reply:
paging is used for the common implementation. e.g, for patient export, the codes read page of patients, and then write these patients' resources of current to-be-exported resource type to COS, and then move to the next page of patients.
But for group export, we don't really use paging for patients, we get all patient in one batch, so just as mentioned in the comments, we simply set current page number to be 2 and the lastpagenum to be 1 to simply sign the CheckPoint and ChunkWriter to finish writing for current resource type.

Reply:
OK - I got it now.  Thanks.